src/main/java/org/opensearch/search/relevance/transformer/kendraintelligentranking/preprocess/TextTokenizer.java

```diff
@@ -26,26 +30,83 @@ public class TextTokenizer {
 
   public List<List<String>> tokenize(List<String> texts) {
+    if (texts == null) {
+      return new ArrayList<>();
+    }
+
     return texts.stream()
         .map(text -> tokenize(text))
         .collect(Collectors.toList());
   }
 
+  /**
+   * Split the input text into tokens, with post-processing to remove stop words, punctuation, etc.
+   * @param text input text
+   * @return list of tokens
+   */
   public List<String> tokenize(String text) {
-    String[] tokens = text.split(SPLIT_PATTERN.pattern());
-    List<String> validTokens = new ArrayList<>();
-    for (String token : tokens) {
-      if (token.length() == 0) {
+    if (text == null) {
+      return new ArrayList<>();
+    }
+
+    final BreakIterator breakIterator = BreakIterator.getWordInstance(Locale.ENGLISH);
+    breakIterator.setText(text);
+
+    List<String> tokens = new ArrayList();
+    int start = breakIterator.first();
+    String currentWord;
+    for (int end = breakIterator.next(); end != BreakIterator.DONE; start = end, end = breakIterator.next()) {
+      currentWord = text.substring(start, end).stripTrailing().toLowerCase(Locale.ENGLISH);
+      if (currentWord.isEmpty()) {
         continue;
       }
-      String lowerCased = token.toLowerCase(Locale.ENGLISH);
-      if (!STOP_WORDS.contains(lowerCased)) {
-        validTokens.add(lowerCased);
+      // Split long words
+      List<String> shortenedTokens = new ArrayList<>();
+      if (currentWord.length() <= MAXIMUM_WORD_LENGTH) {
+        shortenedTokens.add(currentWord);
+      } else {
+        for (int i = 0; i < currentWord.length(); i += MAXIMUM_WORD_LENGTH) {
+          shortenedTokens.add(currentWord.substring(i, Math.min(currentWord.length(), i + MAXIMUM_WORD_LENGTH)));
+        }
+      }
+      // Filter out punctuation, short words, numbers
+      for (String shortenedToken : shortenedTokens) {
+        if (!isWordAllPunctuation(shortenedToken) && !STOP_WORDS.contains(shortenedToken) &&
+            shortenedToken.length() >= MINIMUM_WORD_LENGTH && !isNumeric(shortenedToken)) {
+          String tokenWithInWordPunctuationRemoved = removeInWordPunctuation(shortenedToken);
+          if (!tokenWithInWordPunctuationRemoved.isEmpty()) {
+            tokens.add(tokenWithInWordPunctuationRemoved);
+          }
+        }
       }
     }
-    return validTokens;
+    return tokens;
+  }
+
+  boolean isWordAllPunctuation(final String token) {
+    return (token != null) && ALL_PUNCTUATIONS_REGEX.matcher(token).matches();
+  }
+
+  boolean isNumeric(final String token) {
+    if (token == null) {
+      return false;
+    }
+    try {
+      Double.parseDouble(token);
+    } catch (NumberFormatException e) {
+      return false;
+    }
+    return true;
+  }
+
+  String removeInWordPunctuation(String token) {
+    if (token == null) {
+      return null;
+    }
+    return token.replaceAll(PUNCTUATIONS_REGEX_PATTERN, "");
```

Comment:
Can you compile the regex once into a `static final Pattern`, and replace this with:

```
PUNCTUATIONS_REGEX.matcher(token).replaceAll("");
```

As written, this will compile the regex on every token.

Reply:
Ah! I changed the one above and missed this one.. thanks for catching!